### PR TITLE
Add neural network modules, tests, and test utils

### PR DESCRIPTION
**Pull request by kevinchern**

### CODE REVIEW

File: dwave/plugins/torch/nn.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            self.linear = nn.Linear(din, dout, bias=False)
+
+    def forward(self, x) -> torch.Tensor:
+        """Apply a linear transformation to the input variable `x`.
+
+        Args:
+            x (torch.Tensor): the input tensor.
+
+        Returns:
+            torch.Tensor: the linearly-transformed tensor of `x`.
```
Comment (thisac) [inline dwave/plugins/torch/nn.py]:
```suggestion
        """Apply a linear transformation to the input variable ``x``.

        Args:
            x (torch.Tensor): The input tensor.

        Returns:
            torch.Tensor: The linearly-transformed tensor of ``x``.
```

File: dwave/plugins/torch/nn.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        return self.linear(x)
+
+
+class LinearBlock(nn.Module):
+    @store_config
+    def __init__(self, din, dout, p) -> None:
```
Comment (thisac) [inline dwave/plugins/torch/nn.py]: Missing typehints.

File: dwave/plugins/torch/nn.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+import inspect
+from functools import wraps
+from types import MappingProxyType
+
+import torch
+from torch import nn
+
+
```
Comment (thisac) [inline dwave/plugins/torch/nn.py]: Don't forget to add an `__all__` here. See e.g., [here](https://github.com/dwavesystems/dwave-pytorch-plugin/blob/aa981186b1582c133019bdb7d0311b31f5228113/dwave/plugins/torch/models/boltzmann_machine.py#L46).

File: dwave/plugins/torch/nn.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        linear_1 = nn.Linear(din, dout)
+        linear_2 = nn.Linear(dout, dout)
+        self.block = nn.Sequential(
+            nn.LayerNorm(din),
+            linear_1,
+            nn.Dropout(p),
+            nn.ReLU(),
+            nn.LayerNorm(dout),
+            linear_2,
+        )
```
Comment (thisac) [inline dwave/plugins/torch/nn.py]: Is there a reason for not just doing the following? Since each linear layer is only used once it makes more sense to create them in place, as the other layers.
```suggestion
        self.block = nn.Sequential(
            nn.LayerNorm(din),
            nn.Linear(din, dout),
            nn.Dropout(p),
            nn.ReLU(),
            nn.LayerNorm(dout),
            nn.Linear(dout, dout),
        )
```
or even adding in the skip, and removing it from forward
```suggestion
        self.block = nn.Sequential(
            nn.LayerNorm(din),
            nn.Linear(din, dout),
            nn.Dropout(p),
            nn.ReLU(),
            nn.LayerNorm(dout),
            nn.Linear(dout, dout),
            SkipLinear(din, dout),
        )
```

File: dwave/plugins/torch/nn.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        return x
+
+
+class SkipLinear(nn.Module):
+
+    @store_config
+    def __init__(self, din, dout) -> None:
+        """Applies a linear transformation to the incoming data: :math:`y = xA^T`.
+
+        This module is identity when `din == dout`, otherwise, it is a linear transformation, i.e.,
```
Comment (thisac) [inline dwave/plugins/torch/nn.py]:
```suggestion
        This module is identity when ``din == dout``, otherwise, it is a linear transformation, i.e.,
```

File: dwave/plugins/torch/nn.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        # Use default values if the args/kwargs were not supplied
+        bound.apply_defaults()
+        config = {k: v for k, v in bound.arguments.items() if k != 'self'}
+        config['module_name'] = bound.args[0].__class__.__name__
+        self.config = MappingProxyType(config)
+        fn(self, *args, **kwargs)
+    return wrapper
+
+
+class Identity(nn.Module):
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply dwave/plugins/torch/nn.py]: uh LOL. How did I miss that 😆

File: dwave/plugins/torch/nn.py (hunk 5 above)
Comment (VolodyaCO) [inline dwave/plugins/torch/nn.py]: Is this what makes it a "Skip" linear instead of a plain linear module? I find this very counter-intuitive: that having the same output dimension performs an identity transformation instead of a linear transformation (like an endomorphism).

File: dwave/plugins/torch/nn.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        6. a second linear layer, and, finally,
+        7. a skip connection from initial input to output.
+
+        Args:
+            din (int): Size of each input sample
+            dout (int): Size of each output sample
+            p (float): Dropout probability.
+        """
+        super().__init__()
+        self.skip = SkipLinear(din, dout)
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply dwave/plugins/torch/nn.py]: The input of forward is `x`, if you add skip to the end of sequential, it would be performing the operation `skip(block(x))`, but what is wanted is `skip(x) + block(x)`.

**Review by VolodyaCO — COMMENTED** [review]
Left a question about `SkipLinear`

File: dwave/plugins/torch/nn.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        """
+        sig = inspect.signature(fn)
+        bound = sig.bind(self, *args, **kwargs)
+        bound.apply_defaults()
+
+        config = {k: v for k, v in bound.arguments.items() if v != self}
+        config['module_name'] = self.__class__.__name__
+        self.config = MappingProxyType(config)
+
+        fn(self, *args, **kwargs)
```
Comment (anahitamansouri) [inline dwave/plugins/torch/nn.py]: I was wondering if you should return `fn(...)` or not?

File: releasenotes/notes/add-nn-modules-c29a092140eacbe1.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - Add the python module ``dwave.plugins.torch.nn`` containing commonly-used neural network modules
+    and patterns used to build more complex architectures.
+  - Add ``LinearBlock`` and ``SkipLinear` modules.
```
Comment (anahitamansouri) [inline releasenotes/notes/add-nn-modules-c29a092140eacbe1.yaml]:
```suggestion
  - Add ``LinearBlock`` and ``SkipLinear`` modules.
```

File: dwave/plugins/torch/nn.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        Args:
+            x (torch.Tensor): The input tensor.
+
+        Returns:
+            torch.Tensor: The linearly-transformed tensor of ``x``.
+        """
+        return self.linear(x)
+
+
+class LinearBlock(nn.Module):
```
Comment (anahitamansouri) [inline dwave/plugins/torch/nn.py]: Could you please motivate the architecture? I assume you have been using this for something and you needed this in the plugin. Right?

**Review by anahitamansouri — COMMENTED** [review]
Thanks very much Kevin for the great work. I have left my comments.

**Comment by kevinchern** [comment]
RE @thisac : restructured the file into (sub)modules by pattern-matching pytorch's structure.
RE @VolodyaCO : configs are now stored recursely for nested models :D. reference for resnet added in the docstring
RE @anahitamansouri : fixed typos and added reference for resnet in docstring

File: dwave/plugins/torch/nn/modules/linear.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+            torch.Tensor: The linearly-transformed tensor of ``x``.
+        """
+        return self.linear(x)
+
+
+class LinearBlock(nn.Module):
+    """A linear block consisting of normalizations, linear transformations, dropout, relu, and a skip connection.
+
+    The module is composed of (in order):
+    1. a first layer norm,
```
Comment (thisac) [inline dwave/plugins/torch/nn/modules/linear.py]: I _think_ Sphinx requires the first element in a list to have an empty line above to render properly (in case we render docs).

```suggestion
    The module is composed of (in order):

    1. a first layer norm,
```

File: dwave/plugins/torch/nn/modules/linear.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    7. a skip connection from initial input to output.
+
+    This is based on the `ResNet paper <https://arxiv.org/abs/1512.03385>`.
+
+    Args:
+        din (int): Size of each input sample.
+        dout (int): Size of each output sample.
+        p (float): Dropout probability.
+    """
+    @store_config
```
Comment (thisac) [inline dwave/plugins/torch/nn/modules/linear.py]:
```suggestion
    """

    @store_config
```

File: dwave/plugins/torch/nn/modules/linear.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            nn.LayerNorm(din),
+            linear_1,
+            nn.Dropout(p),
+            nn.ReLU(),
+            nn.LayerNorm(dout),
+            linear_2,
+        )
+
+    def forward(self, x) -> torch.Tensor:
+        """Transforms the input `x` with the modules.
```
[the remark this answers is not in the document]
Comment (thisac) [reply dwave/plugins/torch/nn/modules/linear.py]: Missed this one!

File: tests/test_nn.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+import torch
+from parameterized import parameterized
+
+from dwave.plugins.torch.nn import LinearBlock, SkipLinear, store_config
+from tests.utils import model_probably_good
+
+
+class TestUtils(unittest.TestCase):
+
+    def test_store_config(self):
```
Comment (thisac) [inline tests/test_nn.py]: Should split the simple case and nested case into two separate tests.

File: tests/test_nn.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+        class OuterModel(torch.nn.Module):
+            @store_config
+            def __init__(self, module_1, module_2=None):
+                super().__init__()
+
+        module_1 = InnerModel(a=123, x=5)
+        module_2 = InnerModel(a="second", y="lol")
+        model = OuterModel(module_1, module_2)
+        self.assertDictEqual(dict(model.config),
```
Comment (thisac) [inline tests/test_nn.py]: Does this work well comparing nested `MappingProxyType` objects? I wonder if you'd need to test e.g.,
```python
self.assertDictEqual(dict(model.config[module_1]), {"a": 123, "b": 1:, "x": 5, "y": "hello"}
self.assertDictEqual(dict(model.config[module_2]), {"a": "second", "b": 1:, "x": 4, "y": "lol"}
```
separately as well.

File: tests/test_nn.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        dout = 177
+        model = LinearBlock(din, dout, p)
+        self.assertTrue(model_probably_good(model, (din,), (dout,)))
+
+    def test_SkipLinear(self):
+        din = 33
+        dout = 99
+        model = SkipLinear(din, dout)
+        self.assertTrue(model_probably_good(model, (din,), (dout, )))
+        with self.subTest("Check identity for `din == dout`"):
```
Comment (thisac) [inline tests/test_nn.py]: This seems like it should be a stand-alone test.

File: tests/test_test_utils.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
+import unittest
+
+import torch
+from parameterized import parameterized
+
+from dwave.plugins.torch.nn import store_config
+from tests import utils
+
+
+class TestUtils(unittest.TestCase):
```
Comment (thisac) [inline tests/test_test_utils.py]: Not sure I like the `test_test_utils.py` name. Maybe better to call this something different. Perhaps "helper_functions"? Not sure. Or simply call the file `test_utilities.py` and have it contain both the utility function _and_ it's tests. Or rename the other `test_utils.py` to `test_torch_utils.py`.

File: tests/test_test_utils.py [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+        class MyModel(torch.nn.Module):
+            @store_config
+            def __init__(self, a, b=2, *, x=4, y=5):
+                super().__init__()
+
+            def forward(self, x):
+                return torch.ones(5)
+        model = MyModel(**kwargs)
+        self.assertTrue(utils.has_correct_config(model))
+        self.assertFalse(utils.has_correct_config(torch.nn.Linear(5, 3)))
```
Comment (thisac) [inline tests/test_test_utils.py]: This gets run twice (identically) due to the parameterization. I'd put it as a separate test.

File: tests/utils.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    Args:
+        model (torch.nn.Module): The module to be tested.
+        shape_in (tuple[int, ...]): Input data shape excluding the batch dimension.
+        shape_out (tuple[int, ...]): Output data shape excluding the batch dimension.
+
+    Returns:
+        bool: Indicator for whether the model meets the three conditions above.
+    """
+    bs = 100
+    x = torch.randn((bs, ) + shape_in)
```
Comment (thisac) [inline tests/utils.py]: Never too fond of randomness in tests, if it can be avoided. 🫤

File: tests/utils.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+    Args:
+        x (torch.Tensor): A tensor.
+
+    Returns:
+        bool: Indicator for whether `x` has any zero-valued entries.
+    """
+    return (x == 0).float().any()
+
+
+def bounded_in_plus_minus_one(x: torch.Tensor):
```
Comment (thisac) [inline tests/utils.py]:
```suggestion
def bounded_in_plus_minus_one(x: torch.Tensor) -> bool:
```

File: dwave/plugins/torch/nn/modules/linear.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        din (int): Size of each input sample.
+        dout (int): Size of each output sample.
+        p (float): Dropout probability.
+    """
+    @store_config
+    def __init__(self, din: int, dout: int, p: float) -> None:
+        super().__init__()
+        self._skip = SkipLinear(din, dout)
+        dhid = max(din, dout)
+        self.block = nn.Sequential(
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply dwave/plugins/torch/nn/modules/linear.py]: oh yeah thanks! that's definitely a mistake lol

File: tests/utils.py (hunk 19 above)
Comment (kevinchern) [reply tests/utils.py]: I think hard-coded are more likely to miss bugs than this though. Do u have a suggestion for what hard-coded value would be useful here? (Or alternatives to this test)

Comment (kevinchern) [reply tests/utils.py]: Or maybe this is not the right way to test these modules...

File: tests/utils.py [hunk 22]
```diff
@@ -0,0 +1,123 @@
+from inspect import signature
+
+import torch
+
+
+def model_probably_good(
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply tests/utils.py]: Not sure. I only ever use it as unit "tests", more like sanity checks, for modules

File: tests/test_test_utils.py [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+class TestUtils(unittest.TestCase):
+
+    def test_probably_unconstrained(self):
+        x = torch.randn((1000, 10, 10))
+        self.assertTrue(utils.probably_unconstrained(x))
+
+        # Activate
+        self.assertFalse(utils.probably_unconstrained(x.sigmoid()))
+        self.assertFalse(utils.probably_unconstrained(x.relu()))
+        self.assertFalse(utils.probably_unconstrained(x.tanh()))
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply tests/test_test_utils.py]: Huh, why? Isn't this standard practice?

File: tests/test_nn.py (hunk 16 above)
Comment (kevinchern) [reply tests/test_nn.py]: Why?

File: tests/test_nn.py [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+    def test_store_config(self):
+        # Simple case
+        class MyModel(torch.nn.Module):
+            @store_config
+            def __init__(self, a, b=1, *, x=4, y='hello'):
+                super().__init__()
+
+        model = MyModel(a=123, x=5)
+        self.assertDictEqual(dict(model.config),
+                             {"a": 123, "b": 1, "x": 5, "y": "hello", "module_name": "MyModel"})
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply tests/test_nn.py]: what case do u have in mind?

File: tests/test_nn.py (hunk 16 above)
Comment (thisac) [reply tests/test_nn.py]: Because testing the case when `din == dout` is a separate case to the one where `din != dout`. Having them as separate tests makes it a bit clearer and easier to parse the results. Also, there's very little code deduplication by having it in one test. 

Each test case should optimally be run separately, either as subtests, parameterized tests, or completely stand-alone.

File: tests/test_nn.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (thisac) [reply tests/test_nn.py]: Not necessarily necessary here, but currently you're only testing for `MyModel(a=123, x=5)`. Other cases could be with different types (are only strings/ints allowed?), args instead of kwargs (or combinations thereof). Also, separate tests could be checking that storing a config with empty signature (only a name) works, only with default values, etc.

File: tests/test_nn.py (hunk 16 above)
Comment (thisac) [reply tests/test_nn.py]: Yeah, that makes sense to me.

**Comment by kevinchern** [comment]
@thisac  updated again based on your feedback on best practices. Thanks!!

1) added more tests for the store_config case and separated them as recommended
2) split linear tests as recommended

File: dwave/plugins/torch/nn/modules/linear.py
Comment (thisac) [inline dwave/plugins/torch/nn/modules/linear.py]:
```suggestion
        """Transforms the input ``x`` with the modules.
```

File: tests/test_helper_functions.py [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+    def test_model_probably_good(self):
+        with self.subTest("Model should be good"):
+            class MyModel(torch.nn.Module):
+                @store_config
+                def __init__(self, a, b=2, *, x=4, y=5):
+                    super().__init__()
+
+                def forward(self, x):
+                    return 2*x
+            self.assertTrue(helper_functions.model_probably_good(MyModel("hello"), (500, ), (500,)))
```
Comment (thisac) [inline tests/test_helper_functions.py]: To a point made earlier, this test could be simplified a bit:
```suggestion
        class MyModel(torch.nn.Module):
            @store_config
            def __init__(self, a, b=2, *, x=4, y=5):
                super().__init__()

        with self.subTest("Model should be good"):
            MyModel.forward = lambda self, x: 2 * x
            self.assertTrue(helper_functions.model_probably_good(MyModel("hello"), (500, ), (500,)))
```
but I'll leave this just as an option. Up to you whether you prefer this or the former.

File: tests/test_test_utils.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (kevinchern) [reply tests/test_test_utils.py]: sry missed this one. for posterity: wouldn't this reasoning imply every test can only have a single assertion?
the test is for one behaviour on three input values of same shape and dtype, which seems reasonable to be in a single test

File: tests/test_helper_functions.py [hunk 26]
```diff
@@ -0,0 +1,148 @@
+import unittest
+
+import torch
+from parameterized import parameterized
+
+from dwave.plugins.torch.nn import store_config
+from tests import helper_functions
+
+
+class TestUtils(unittest.TestCase):
```
Comment (thisac) [inline tests/test_helper_functions.py]:
```suggestion
class TestHelperFunctions(unittest.TestCase):
```
